shuffle-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/ucx/UCX.scala

```diff
@@ -345,6 +362,235 @@ class UCX(executor: BlockManagerId, rapidsConf: RapidsConf) extends AutoCloseabl
     })
   }
 
+  /**
+   * This represents the mapping between an Active Message Id and the callback that should
+   * be triggered when a message is received.
+   *
+   * There are two types of Active Messages we care about: requests and responses.
+   *
+   * For requests:
+   *   - `activeMessageId` for requests is the value of the `RequestType` enum, and it is
+   *   set once when the `RapidsShuffleServer` is initialized, and no new request handlers
+   *   are established.
+   *
+   *   - The Active Message header is handed to the request handler, via the transaction.
+   *   The request handler needs to echo the header back for the response handler on the
+   *   other side of the request.
+   *
+   *   - On a request, a callback is instantiated using `requestCallbackGen`, which creates
+   *   a transaction each time. This is one way to handle several requests inbound to a server.
+   *
+   * For responses:
+   *   - `activeMessageId` for responses is the value of the `RequestType` enum with an extra
+   *   bit flipped (see `UCXConnection.composeResponseAmId`). These are also set once, as requests
+   *   are sent out.
+   *
+   *   - The Active Message header is used to pick the correct callback to call. In this case
+   *   there could be serveral expected responses, for a single response `activeMessageId`, so the
+   *   server echoes back our header so we can invoke the correct response callback.
+   *
+   *   - Each response received at the response activeMessageId, will be demuxed using the header:
+   *   responseActiveMessageId1 -> [callbackForHeader1, callbackForHeader2, ..., callbackForHeaderN]
+   */
+  private class ActiveMessageRegistration(val activeMessageId: Int) {
```

Comment:
This feels like a class that is trying to be two different things at once, which makes me wonder if there should be an interface like this:
```scala
trait ActiveMessageRegistration {
  val activeMessageId
  def getCallback(header: Long): UCXAmCallback
}
```
And then there's two types that implement this interface:
- a request registration that takes the callback generator as a constructor argument
- a response registration that provides the `addResponseActiveMessageHandler` method

Not must-fix, could also be a followup PR to clean it up, but all the "make sure I'm the right type of instance" checks in the methods makes me wonder if it would be cleaner to make the registration types explicit.